README.md
[![pipeline](https://gitlab.com/d-e-s-o/time-util/badges/master/pipeline.svg)](https://gitlab.com/d-e-s-o/time-util/commits/master)
[![coverage](https://gitlab.com/d-e-s-o/time-util/badges/master/coverage.svg)](https://gitlab.com/d-e-s-o/time-util/-/jobs/artifacts/master/file/kcov/kcov-merged/index.html?job=coverage:kcov)
[![crates.io](https://img.shields.io/crates/v/time-util.svg)](https://crates.io/crates/time-util)
[![Docs](https://docs.rs/time-util/badge.svg)](https://docs.rs/time-util)
[![rustc](https://img.shields.io/badge/rustc-1.39+-blue.svg)](https://blog.rust-lang.org/2019/11/07/Rust-1.39.0.html)

time-util
=========

- [Documentation][docs-rs]

A crate providing miscellaneous functionality for working with time
stamps.


[docs-rs]: https://docs.rs/crate/time-util


CHANGELOG.md
0.3.4
-----
- Add support for printing a time as ISO 8601 date
- Fixed `docs.rs` documentation build


0.3.3
-----
- Added support for printing and serializing time as RFC 3339 time stamp
  with nanosecond resolution


0.3.2
-----
- Added missing use declarations when feature `chrono-tz` is enabled
- Enabled CI pipeline comprising building, testing, and linting of the
  project


0.3.1
-----
- Added support for printing a time as RFC 3339 time stamp


0.3.0
-----
- Use `chrono-tz` for time-zone aware conversions
  - Removed custom time-zone logic


0.2.0
-----
- Added support for parsing a time stamp from a date string
- Added support for serializing a time stamp for a given time zone
- Added support for rudimentary time calculations
- Split crate into three modules: `math`, `parse`, and `serde`
  - Each has to be opted-in with the corresponding feature flag


0.1.0
-----
- Initial release


src/lib.rs
// Copyright (C) 2019-2021 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

//! A crate for more or less frequently used time parsing and formatting
//! functionality.
//!
//! The crate revolves around the `std::time::SystemTime` type in that
//! we attempt to convert into that or use this as the base to convert
//! from. We treat such a time as having no associated time zone. Think
//! of it as being in UTC.

#[allow(clippy::let_and_return)]

#[cfg(feature = "math")]
mod math;
#[cfg(any(test, feature = "chrono"))]
mod parse;
#[cfg(any(test, feature = "chrono"))]
mod print;
#[cfg(any(test, all(feature = "chrono", feature = "serde")))]
mod serde;

// We treat chrono-tz as optional on top of chrono.
#[cfg(not(any(feature = "math", feature = "chrono", feature = "serde")))]
compile_error!("Please specify one of the features: math, chrono, or serde");

#[cfg(feature = "math")]
pub use crate::math::{
  days_back,
  days_back_from,
  next_day,
  tomorrow,
};

#[cfg(feature = "chrono")]
pub use crate::parse::{
  parse_system_time_from_date_str,
  parse_system_time_from_str,
};

#[cfg(feature = "chrono")]
pub use crate::print::{
  print_system_time_to_iso8601_date,
  print_system_time_to_rfc3339,
  print_system_time_to_rfc3339_with_nanos,
};

#[cfg(all(feature = "chrono", feature = "serde"))]
pub use crate::serde::{
  optional_system_time_from_str,
  optional_system_time_to_rfc3339,
  optional_system_time_to_rfc3339_with_nanos,
  system_time_from_date_str,
  system_time_from_millis,
  system_time_from_secs,
  system_time_from_str,
  system_time_to_millis,
  system_time_to_rfc3339,
  system_time_to_rfc3339_with_nanos,
};

#[cfg(all(feature = "chrono", feature = "chrono-tz", feature = "serde"))]
pub use crate::serde::{
  system_time_from_millis_in_new_york,
  system_time_to_millis_in_new_york,
};


src/math.rs
// Copyright (C) 2020 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::time::Duration;
use std::time::SystemTime;
use std::time::UNIX_EPOCH;

use math_util::round_up;


/// The number of seconds in a day.
const DAY_SECS: u32 = 24 * 60 * 60;


fn next_day_duration(now: SystemTime) -> Duration {
  // `UNIX_EPOCH` is the first time stamp that can be represented, so
  // there is no way `SystemTime::duration_since` can ever fail with it
  // as a parameter.
  let duration = now.duration_since(UNIX_EPOCH).unwrap();
  let next_day = round_up(duration.as_secs(), DAY_SECS.into());
  let duration = Duration::from_secs(next_day);
  duration
}

/// Calculate the time stamp representing the next day of the given time
/// stamp.
///
/// Note that currently a time stamp marking precisely midnight will not
/// advance to the next day.
// TODO: We should fix this behavior.
pub fn next_day(now: SystemTime) -> SystemTime {
  let duration = next_day_duration(now);
  UNIX_EPOCH + duration
}

pub fn days_back_from(now: SystemTime, count: u32) -> SystemTime {
  let duration = next_day_duration(now) - Duration::from_secs(DAY_SECS.into()) * (count + 1);
  UNIX_EPOCH + duration
}

pub fn days_back(count: u32) -> SystemTime {
  days_back_from(SystemTime::now(), count)
}

/// Calculate a `SystemTime` representing 0:00:00 (i.e., the first
/// second of) the next day.
pub fn tomorrow() -> SystemTime {
  next_day(SystemTime::now())
}


#[cfg(test)]
pub mod tests {
  use super::*;

  use crate::parse::parse_system_time_from_str;


  #[test]
  fn calculate_next_day() {
    let now = parse_system_time_from_str("2020-02-07T13:00:00Z").unwrap();
    let tomorrow = parse_system_time_from_str("2020-02-08T00:00:00Z").unwrap();
    assert_eq!(next_day(now), tomorrow);

    let now = parse_system_time_from_str("2019-03-31T23:59:59Z").unwrap();
    let tomorrow = parse_system_time_from_str("2019-04-01T00:00:00Z").unwrap();
    assert_eq!(next_day(now), tomorrow);
  }

  #[test]
  fn calculate_prev_days() {
    let now = parse_system_time_from_str("2020-02-07T09:00:00Z").unwrap();
    let yesterday = parse_system_time_from_str("2020-02-06T00:00:00Z").unwrap();
    assert_eq!(days_back_from(now, 1), yesterday);

    let now = parse_system_time_from_str("2020-02-02T23:59:59Z").unwrap();
    let five_ago = parse_system_time_from_str("2020-01-28T00:00:00Z").unwrap();
    assert_eq!(days_back_from(now, 5), five_ago);
  }
}


src/parse.rs
// Copyright (C) 2020-2021 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::time::SystemTime;

use chrono::naive::NaiveDate;
use chrono::offset::FixedOffset;
use chrono::offset::TimeZone as _;
use chrono::DateTime;
use chrono::ParseError;


type DateFn = fn(&str) -> Result<DateTime<FixedOffset>, ParseError>;


/// The list of time stamp formats we support.
pub(crate) const TIME_PARSE_FNS: [DateFn; 3] = [
  |s| FixedOffset::east(0).datetime_from_str(s, "%Y-%m-%dT%H:%M:%S%.fZ"),
  |s| FixedOffset::east(0).datetime_from_str(s, "%Y-%m-%dT%H:%M:%SZ"),
  |s| DateTime::parse_from_str(s, "%Y-%m-%dT%H:%M:%S%.f%z"),
];

pub(crate) const DATE_PARSE_FNS: [DateFn; 1] = [|s| {
  NaiveDate::parse_from_str(s, "%Y-%m-%d")
    .map(|date| DateTime::from_utc(date.and_hms(0, 0, 0), FixedOffset::east(0)))
}];


/// Parse a `SystemTime` from a string using any of the provided parsing
/// functions.
pub(crate) fn parse_system_time_from_str_impl(
  time: &str,
  parse_fns: &[DateFn],
) -> Option<SystemTime> {
  for parse_fn in parse_fns {
    // Ideally we would want to only continue in case of
    // ParseErrorKind::Invalid. However, that member is private...
    let datetime = match parse_fn(&time) {
      Ok(datetime) => datetime,
      Err(_) => continue,
    };
    return Some(SystemTime::from(
      datetime.with_timezone(&FixedOffset::east(0)),
    ))
  }
  None
}


/// Parse a `SystemTime` from a string.
pub fn parse_system_time_from_str(time: &str) -> Option<SystemTime> {
  parse_system_time_from_str_impl(&time, &TIME_PARSE_FNS)
}


/// Parse a `SystemTime` from a date string.
pub fn parse_system_time_from_date_str(time: &str) -> Option<SystemTime> {
  parse_system_time_from_str_impl(&time, &DATE_PARSE_FNS)
}


#[cfg(test)]
mod tests {
  use super::*;

  use std::time::Duration;
  use std::time::UNIX_EPOCH;


  #[test]
  fn parse_time() {
    let time = parse_system_time_from_str("2018-04-01T12:00:00.000Z").unwrap();
    let expected = UNIX_EPOCH + Duration::from_secs(1_522_584_000);
    assert_eq!(time, expected)
  }

  /// Check that parsing an unexpected time fails.
  #[test]
  fn parse_time_fail() {
    let result = parse_system_time_from_str("2018-04-01TXX:00:00.000");
    assert_eq!(result, None)
  }

  #[test]
  fn parse_date() {
    let time = parse_system_time_from_date_str("2019-08-01").unwrap();
    let expected = UNIX_EPOCH + Duration::from_secs(1_564_617_600);
    assert_eq!(time, expected)
  }
}


src/print.rs
// Copyright (C) 2021 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::time::SystemTime;

use chrono::offset::Utc;
use chrono::DateTime;
use chrono::SecondsFormat;


/// Print a `SystemTime` as a RFC3339 time stamp.
pub fn print_system_time_to_rfc3339(time: &SystemTime) -> String {
  DateTime::<Utc>::from(*time).to_rfc3339_opts(SecondsFormat::Millis, true)
}


/// Print a `SystemTime` as a RFC3339 time stamp.
pub fn print_system_time_to_rfc3339_with_nanos(time: &SystemTime) -> String {
  // Rust's `SystemTime` internally work with nano seconds and so by
  // doing the same we hope to have no loss of information.
  DateTime::<Utc>::from(*time).to_rfc3339_opts(SecondsFormat::Nanos, true)
}


/// Print a `SystemTime` as a RFC3339 time stamp.
pub fn print_system_time_to_iso8601_date(time: &SystemTime) -> String {
  DateTime::<Utc>::from(*time).format("%Y-%m-%d").to_string()
}


#[cfg(test)]
mod tests {
  use super::*;

  use crate::parse::parse_system_time_from_str;


  /// Check that we can format a `SystemTime` as a RFC3339 time stamp.
  #[test]
  fn print_rfc3339() {
    let string = "2018-04-01T12:04:17.050Z";
    let time = parse_system_time_from_str(string).unwrap();
    let result = print_system_time_to_rfc3339(&time);
    assert_eq!(result, string)
  }


  /// Check that we can format a `SystemTime` as a RFC3339 time stamp
  /// with nanosecond resolution.
  #[test]
  fn print_rfc3339_with_nanos() {
    let string = "2018-04-01T12:04:17.050937231Z";
    let time = parse_system_time_from_str(string).unwrap();
    let result = print_system_time_to_rfc3339_with_nanos(&time);
    assert_eq!(result, string)
  }


  /// Check that we can format a `SystemTime` as an ISO 8601 date.
  #[test]
  fn print_iso8601_date() {
    let string = "2018-05-07T23:59:59Z";
    let time = parse_system_time_from_str(string).unwrap();
    let result = print_system_time_to_iso8601_date(&time);
    assert_eq!(result, "2018-05-07");

    let string = "2018-05-07T00:00:00Z";
    let time = parse_system_time_from_str(string).unwrap();
    let result = print_system_time_to_iso8601_date(&time);
    assert_eq!(result, "2018-05-07");
  }
}


src/timezone.rs
// Copyright (C) 2020 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::time::Duration;
use std::time::SystemTime;


/// An enumeration describing a timezone offset.
pub enum Offset {
  /// There is no offset, i.e., the time zone is UTC.
  None,
  /// An offset from UTC (in seconds) in the western hemisphere.
  ///
  /// E.g., a value of `1 * 60 * 60` maps to UTC-01:00
  West(u16),
  /// An offset from UTC (in seconds) in the eastern hemisphere.
  ///
  /// E.g., a value of `1 * 60 * 60` maps to UTC+01:00
  East(u16),
}


/// A trait representing a timezone and operations defined for it.
pub trait TimeZone {
  const OFFSET: Offset;

  /// Correct a system time by adding our offset.
  fn add(time: SystemTime) -> SystemTime {
    match Self::OFFSET {
      Offset::None => time,
      Offset::West(offset) => time - Duration::from_secs(offset.into()),
      Offset::East(offset) => time + Duration::from_secs(offset.into()),
    }
  }

  /// Correct a system time by subtracting our offset.
  fn sub(time: SystemTime) -> SystemTime {
    match Self::OFFSET {
      Offset::None => time,
      Offset::West(offset) => time + Duration::from_secs(offset.into()),
      Offset::East(offset) => time - Duration::from_secs(offset.into()),
    }
  }
}


/// The UTC timezone.
pub struct UTC {}

impl TimeZone for UTC {
  const OFFSET: Offset = Offset::None;
}


/// The eastern standard time zone.
pub struct EST {}

impl TimeZone for EST {
  const OFFSET: Offset = Offset::West(5 * 60 * 60);
}


#[cfg(test)]
mod tests {
  use super::*;

  use crate::parse::parse_system_time_from_str;


  #[test]
  fn time_zone_correction() {
    let est_time = parse_system_time_from_str("2018-04-01T08:00:37.000-05:00").unwrap();
    let utc_time = parse_system_time_from_str("2018-04-01T08:00:37.000-00:00").unwrap();
    let expected = parse_system_time_from_str("2018-04-01T08:00:37.000Z").unwrap();

    assert_eq!(EST::add(est_time), expected);
    assert_eq!(UTC::add(utc_time), expected);
  }
}


src/serde.rs
// Copyright (C) 2020-2021 Daniel Mueller <deso@posteo.net>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::time::Duration;
use std::time::SystemTime;
use std::time::UNIX_EPOCH;

#[cfg(feature = "chrono-tz")]
use chrono::{
  offset::TimeZone as _,
  DateTime,
  Utc,
};
#[cfg(feature = "chrono-tz")]
use chrono_tz::America::New_York;

use serde::de::Deserializer;
use serde::de::Error;
use serde::de::Unexpected;
use serde::ser::Serializer;
use serde::Deserialize;

use crate::parse::parse_system_time_from_str_impl;
use crate::parse::DATE_PARSE_FNS;
use crate::parse::TIME_PARSE_FNS;
use crate::print::print_system_time_to_rfc3339;
use crate::print::print_system_time_to_rfc3339_with_nanos;


/// Deserialize a time stamp as a `SystemTime`.
pub fn system_time_from_str<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
  D: Deserializer<'de>,
{
  let time = String::deserialize(deserializer)?;
  parse_system_time_from_str_impl(&time, &TIME_PARSE_FNS)
    .ok_or_else(|| Error::invalid_value(Unexpected::Str(&time), &"a time stamp string"))
}


/// Deserialize an optional time stamp.
pub fn optional_system_time_from_str<'de, D>(
  deserializer: D,
) -> Result<Option<SystemTime>, D::Error>
where
  D: Deserializer<'de>,
{
  match Option::<String>::deserialize(deserializer)? {
    Some(time) => parse_system_time_from_str_impl(&time, &TIME_PARSE_FNS)
      .ok_or_else(|| Error::invalid_value(Unexpected::Str(&time), &"an optional time stamp string"))
      .map(Option::Some),
    None => Ok(None),
  }
}


/// Deserialize a `SystemTime` from a date.
pub fn system_time_from_date_str<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
  D: Deserializer<'de>,
{
  let date = String::deserialize(deserializer)?;
  parse_system_time_from_str_impl(&date, &DATE_PARSE_FNS)
    .ok_or_else(|| Error::invalid_value(Unexpected::Str(&date), &"a date string"))
}


/// Deserialize a `SystemTime` from a UNIX time stamp.
pub fn system_time_from_secs<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
  D: Deserializer<'de>,
{
  let seconds = u64::deserialize(deserializer)?;
  let time = UNIX_EPOCH + Duration::new(seconds, 0);
  Ok(time)
}


/// Deserialize a `SystemTime` from a timestamp containing the
/// milliseconds since 1970-01-01.
pub fn system_time_from_millis<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
  D: Deserializer<'de>,
{
  let ms = u64::deserialize(deserializer)?;
  let time = UNIX_EPOCH + Duration::from_millis(ms);
  Ok(time)
}


/// Deserialize a `SystemTime` from a timestamp containing the
/// milliseconds since 1970-01-01 in the New York time zone.
#[cfg(feature = "chrono-tz")]
pub fn system_time_from_millis_in_new_york<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
  D: Deserializer<'de>,
{
  let time = system_time_from_millis(deserializer)?;
  let naive_time = DateTime::<Utc>::from(time).naive_local();
  let ny_time = New_York.from_utc_datetime(&naive_time);
  let utc_time = Utc.from_local_datetime(&ny_time.naive_local()).unwrap();

  Ok(SystemTime::from(utc_time))
}


/// Serialize a `SystemTime` into a RFC3339 time stamp.
pub fn system_time_to_rfc3339<S>(time: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  let string = print_system_time_to_rfc3339(time);
  serializer.serialize_str(&string)
}


/// Serialize a `SystemTime` into a RFC3339 time stamp.
pub fn system_time_to_rfc3339_with_nanos<S>(
  time: &SystemTime,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  let string = print_system_time_to_rfc3339_with_nanos(time);
  serializer.serialize_str(&string)
}


/// Serialize an optional `SystemTime` into a RFC3339 time stamp.
pub fn optional_system_time_to_rfc3339<S>(
  time: &Option<SystemTime>,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  match time {
    Some(time) => system_time_to_rfc3339(time, serializer),
    None => serializer.serialize_none(),
  }
}


/// Serialize an optional `SystemTime` into a RFC3339 time stamp.
pub fn optional_system_time_to_rfc3339_with_nanos<S>(
  time: &Option<SystemTime>,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  match time {
    Some(time) => system_time_to_rfc3339_with_nanos(time, serializer),
    None => serializer.serialize_none(),
  }
}


/// Serialize a `SystemTime` into a timestamp containing the
/// milliseconds since 1970-01-01.
pub fn system_time_to_millis<S>(time: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  // It should be safe to unwrap here given that there is absolutely no
  // way for a time stamp to ever point to a time before `UNIX_EPOCH`
  // and that the only (documented) error case for `duration_since`.
  let millis = time.duration_since(UNIX_EPOCH).unwrap().as_millis();
  serializer.serialize_u128(millis)
}


/// Serialize a `SystemTime` into a timestamp containing the
/// milliseconds since 1970-01-01 in New York.
#[cfg(feature = "chrono-tz")]
pub fn system_time_to_millis_in_new_york<S>(
  time: &SystemTime,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  let utc_time = DateTime::<Utc>::from(*time);
  let ny_time = New_York.from_local_datetime(&utc_time.naive_utc()).unwrap();
  system_time_to_millis(&SystemTime::from(ny_time), serializer)
}


#[cfg(test)]
mod tests {
  use super::*;

  use std::time::SystemTime;

  use serde::Deserialize;
  use serde::Serialize;
  use serde_json::from_str as from_json;
  use serde_json::to_string as to_json;

  #[cfg(feature = "chrono-tz")]
  use crate::parse::parse_system_time_from_str;


  #[derive(Debug, Deserialize)]
  struct Time {
    #[serde(deserialize_with = "system_time_from_str")]
    time: SystemTime,
  }

  #[test]
  fn deserialize_system_time_from_str() {
    let times = [
      r#"{"time": "2018-04-01T12:00:00Z"}"#,
      r#"{"time": "2018-04-01T12:00:00.000Z"}"#,
      r#"{"time": "2018-04-01T08:00:00.000-04:00"}"#,
    ];

    for json in &times {
      let time = from_json::<Time>(json).unwrap();
      assert_eq!(time.time, UNIX_EPOCH + Duration::from_secs(1_522_584_000));
    }
  }

  #[derive(Debug, Deserialize)]
  struct Date {
    #[serde(deserialize_with = "system_time_from_date_str")]
    date: SystemTime,
  }

  #[test]
  fn deserialize_system_time_from_date_str() {
    let dates = [r#"{"date": "2019-08-01"}"#];

    for json in &dates {
      let date = from_json::<Date>(json).unwrap();
      assert_eq!(date.date, UNIX_EPOCH + Duration::from_secs(1_564_617_600));
    }
  }


  #[derive(Debug, Deserialize, Serialize)]
  struct OtherTime {
    #[serde(
      deserialize_with = "system_time_from_secs",
      serialize_with = "system_time_to_rfc3339",
    )]
    time: SystemTime,
  }

  #[test]
  fn deserialize_system_time_from_secs() {
    let time = from_json::<OtherTime>(r#"{"time": 1544129220}"#).unwrap();
    assert_eq!(time.time, UNIX_EPOCH + Duration::from_secs(1_544_129_220));
  }

  #[test]
  fn serialize_system_time_to_rfc3339() {
    let time = OtherTime {
      time: UNIX_EPOCH + Duration::from_secs(1_544_129_220),
    };
    let json = to_json(&time).unwrap();
    assert_eq!(json, r#"{"time":"2018-12-06T20:47:00.000Z"}"#);
  }

  #[derive(Debug, Deserialize, Serialize)]
  struct MsTime {
    #[serde(
      deserialize_with = "system_time_from_millis",
      serialize_with = "system_time_to_rfc3339",
    )]
    time: SystemTime,
  }

  #[test]
  fn deserialize_system_time_from_millis() {
    let time = from_json::<MsTime>(r#"{"time": 1517461200000}"#).unwrap();
    assert_eq!(time.time, UNIX_EPOCH + Duration::from_secs(1_517_461_200));
  }


  #[derive(Debug, Deserialize, Serialize)]
  #[cfg(feature = "chrono-tz")]
  struct MsTimeNY {
    #[serde(
      deserialize_with = "system_time_from_millis_in_new_york",
      serialize_with = "system_time_to_millis_in_new_york",
    )]
    time: SystemTime,
  }

  #[test]
  #[cfg(feature = "chrono-tz")]
  fn deserialize_serialize_system_time_millis_in_new_york() {
    // This time stamp represents 2018-02-01T00:00:00-05:00:
    // $ date --date='2018-02-01T00:00:00-05:00' +'%s'
    let time = from_json::<MsTimeNY>(r#"{"time": 1517461200000}"#).unwrap();
    let expected = parse_system_time_from_str("2018-02-01T00:00:00.000Z").unwrap();
    assert_eq!(time.time, expected);

    let json = to_json::<MsTimeNY>(&time).unwrap();
    let time = from_json::<MsTimeNY>(&json).unwrap();
    assert_eq!(time.time, expected);
  }

  #[test]
  #[cfg(feature = "chrono-tz")]
  fn deserialize_serialize_system_time_millis_in_new_york_daylight_savings() {
    let time = from_json::<MsTimeNY>(r#"{"time": 1599537600000}"#).unwrap();
    let expected = parse_system_time_from_str("2020-09-08T00:00:00.000Z").unwrap();
    assert_eq!(time.time, expected);

    let json = to_json::<MsTimeNY>(&time).unwrap();
    let time = from_json::<MsTimeNY>(&json).unwrap();
    assert_eq!(time.time, expected);
  }
}
